// @logicflow/extension/es/components/dnd-panel/index.js
var DndPanel = /** @class */ (function () {
    function DndPanel(_a) {
        var _this = this;
        var lf = _a.lf;
        this.lf = lf;
        this.lf.setPatternItems = function (shapeList) {
            _this.shapeList = shapeList;
        };
    }
    DndPanel.prototype.render = function (lf, domContainer) {
        var _this = this;
        if (this.panelEl) {
            domContainer.removeChild(this.panelEl);
        }
        if (!this.shapeList || this.shapeList.length === 0)
            return;
        this.panelEl = document.createElement('div');
        this.panelEl.className = 'lf-dndpanel';
        this.shapeList.forEach(function (shapeItem) {
            _this.panelEl.appendChild(_this.createDndItem(shapeItem));
        });
        domContainer.appendChild(this.panelEl);
    };
    DndPanel.prototype.createDndItem = function (shapeItem) {
        var _this = this;
        var el = document.createElement('div');
        el.className = shapeItem.className ? "lf-dnd-item " + shapeItem.className : 'lf-dnd-item';
        var shape = document.createElement('div');
        shape.className = 'lf-dnd-shape';
        if (shapeItem.icon) {
            shape.style.backgroundImage = "url(" + shapeItem.icon + ")";
        }
        el.appendChild(shape);
        if (shapeItem.label) {
            var text = document.createElement('div');
            text.innerText = shapeItem.label;
            text.className = 'lf-dnd-text';
            el.appendChild(text);
        }
        el.onmousedown = function () {
            if (shapeItem.type) {
                _this.lf.dnd.startDrag({
                    type: shapeItem.type,
                    properties: shapeItem.properties,
                    text: shapeItem.text,
                });
            }
            if (shapeItem.callback) {
                shapeItem.callback();
            }
        };
        return el;
    };
    DndPanel.pluginName = 'DndPanel';
    return DndPanel;
}());
export { DndPanel, };
