// @logicflow/extension/cjs/NodeResize/Node/RectResize.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@logicflow/core");
var ControlGroup_1 = require("../Control/ControlGroup");
var Rect_1 = require("../BasicShape/Rect");
var RectResizeModel = /** @class */ (function (_super) {
    __extends(RectResizeModel, _super);
    function RectResizeModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RectResizeModel.prototype.setAttributes = function () {
        // @ts-ignore
        var nodeSize = this.properties.nodeSize;
        if (nodeSize) {
            this.width = nodeSize.width;
            this.height = nodeSize.height;
        }
    };
    return RectResizeModel;
}(core_1.RectNodeModel));
var RectResizeView = /** @class */ (function (_super) {
    __extends(RectResizeView, _super);
    function RectResizeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RectResizeView.prototype.getControlGroup = function (attributes) {
        var _a = this.props, model = _a.model, graphModel = _a.graphModel;
        return (core_1.h(ControlGroup_1.default, __assign({}, attributes, { nodeModel: model, graphModel: graphModel })));
    };
    // getResizeShape绘制图形，功能等同于基础矩形的getShape功能，可以通过复写此方法，进行节点自定义
    RectResizeView.prototype.getResizeShape = function (arrt) {
        return core_1.h("g", null,
            core_1.h(Rect_1.default, __assign({}, arrt)));
    };
    RectResizeView.prototype.getShape = function () {
        var attributes = _super.prototype.getAttributes.call(this);
        var isSelected = this.props.model.isSelected;
        return (core_1.h("g", null,
            this.getResizeShape(attributes),
            isSelected ? this.getControlGroup(attributes) : ''));
    };
    return RectResizeView;
}(core_1.RectNode));
var RectResize = {
    type: 'rect',
    view: RectResizeView,
    model: RectResizeModel,
};
exports.default = RectResize;
