// @logicflow/extension/cjs/components/mini-map/index.js
"use strict";
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MiniMap = void 0;
var MiniMap = {
    pluginName: 'minimap',
    __lf: null,
    __container: null,
    __miniMapWrap: null,
    __miniMapContainer: null,
    __lfMap: null,
    __viewport: null,
    __width: 150,
    __height: 220,
    __miniMapWidth: 450,
    __miniMapHeight: 660,
    __isShow: false,
    __viewPortTop: 0,
    __viewPortLeft: 0,
    __startPosition: null,
    __viewPortScale: 1,
    __viewPortWidth: 150,
    __viewPortHeight: 75,
    __resetDataX: 0,
    __resetDataY: 0,
    __LogicFlow: null,
    __disabledPlugins: ['minimap', 'control', 'selection-select'],
    install: function (lf, LogicFlow) {
        MiniMap.__lf = lf;
        MiniMap.__miniMapWidth = lf.width;
        MiniMap.__miniMapHeight = (lf.width * 220) / 150;
        MiniMap.__LogicFlow = LogicFlow;
        this.__init();
    },
    init: function (option) {
        this.__disabledPlugins = this.__disabledPlugins.concat(option.disabledPlugins || []);
    },
    render: function (lf, container) {
        MiniMap.__container = container;
        var events = [
            'node:add',
            'node:delete',
            'edge:add',
            'edge:delete',
            'node:drop',
            'blank:drop',
        ];
        events.forEach(function (eventName) {
            MiniMap.__lf.on(eventName, function () {
                MiniMap.__setView();
            });
        });
    },
    show: function (leftPosition, topPosition) {
        MiniMap.__setView();
        if (!MiniMap.__isShow) {
            MiniMap.__createMiniMap(leftPosition, topPosition);
        }
        MiniMap.__isShow = true;
    },
    hide: function () {
        if (MiniMap.__isShow) {
            MiniMap.__removeMiniMap();
        }
        MiniMap.__isShow = false;
    },
    __init: function () {
        var miniMapWrap = document.createElement('div');
        miniMapWrap.className = 'lf-mini-map-graph';
        miniMapWrap.style.width = MiniMap.__width + "px";
        miniMapWrap.style.height = MiniMap.__height + "px";
        MiniMap.__lfMap = new MiniMap.__LogicFlow({
            width: MiniMap.__lf.width,
            height: (MiniMap.__lf.width * 220) / 150,
            container: miniMapWrap,
            isSilentMode: true,
            stopZoomGraph: true,
            stopScrollGraph: true,
            stopMoveGraph: true,
            hideAnchors: true,
            hoverOutline: false,
            disabledPlugins: MiniMap.__disabledPlugins,
        });
        // minimap中禁用adapter。
        MiniMap.__lfMap.adapterIn = function (a) { return a; };
        MiniMap.__lfMap.adapterOut = function (a) { return a; };
        MiniMap.__miniMapWrap = miniMapWrap;
        MiniMap.__createViewPort();
    },
    __createMiniMap: function (left, top) {
        var miniMapContainer = document.createElement('div');
        var miniMapWrap = MiniMap.__miniMapWrap;
        miniMapContainer.appendChild(miniMapWrap);
        if (typeof left !== 'undefined' && typeof top !== 'undefined') {
            miniMapContainer.style.left = left + "px";
            miniMapContainer.style.top = top + "px";
        }
        miniMapContainer.style.position = 'absolute';
        miniMapContainer.className = 'lf-mini-map';
        MiniMap.__container.appendChild(miniMapContainer);
        MiniMap.__miniMapWrap.appendChild(MiniMap.__viewport);
        var header = document.createElement('div');
        header.className = 'lf-mini-map-header';
        header.innerText = '导航';
        miniMapContainer.appendChild(header);
        var close = document.createElement('span');
        close.className = 'lf-mini-map-close';
        close.addEventListener('click', MiniMap.hide);
        miniMapContainer.appendChild(close);
        MiniMap.__miniMapContainer = miniMapContainer;
    },
    __removeMiniMap: function () {
        MiniMap.__container.removeChild(MiniMap.__miniMapContainer);
    },
    /**
     * 计算所有图形一起，占领的区域范围。
     * @param data
     */
    __getBounds: function (data) {
        var left = 0;
        var right = MiniMap.__miniMapWidth;
        var top = 0;
        var bottom = MiniMap.__miniMapHeight;
        var nodes = data.nodes;
        if (nodes && nodes.length > 0) {
            // 因为获取的节点不知道真实的宽高，这里需要补充一点数值
            nodes.forEach(function (_a) {
                var x = _a.x, y = _a.y, _b = _a.width, width = _b === void 0 ? 200 : _b, _c = _a.height, height = _c === void 0 ? 200 : _c;
                var nodeLeft = x - width / 2;
                var noderight = x + width / 2;
                var nodeTop = y - height / 2;
                var nodeBottom = y + height / 2;
                left = nodeLeft < left ? nodeLeft : left;
                right = noderight > right ? noderight : right;
                top = nodeTop < top ? nodeTop : top;
                bottom = nodeBottom > bottom ? nodeBottom : bottom;
            });
        }
        return {
            left: left,
            top: top,
            bottom: bottom,
            right: right,
        };
    },
    /**
     * 将负值的平移转换为正值。
     * 保证渲染的时候，minimap能完全展示。
     * 获取将画布所有元素平移到0，0开始时，所有节点数据
     */
    __resetData: function (data) {
        var nodes = data.nodes, edges = data.edges;
        var left = 0;
        var top = 0;
        if (nodes && nodes.length > 0) {
            // 因为获取的节点不知道真实的宽高，这里需要补充一点数值
            nodes.forEach(function (_a) {
                var x = _a.x, y = _a.y, _b = _a.width, width = _b === void 0 ? 200 : _b, _c = _a.height, height = _c === void 0 ? 200 : _c;
                var nodeLeft = x - width / 2;
                var nodeTop = y - height / 2;
                left = nodeLeft < left ? nodeLeft : left;
                top = nodeTop < top ? nodeTop : top;
            });
            if (left < 0 || top < 0) {
                MiniMap.__resetDataX = left;
                MiniMap.__resetDataY = top;
                nodes.forEach(function (node) {
                    node.x = node.x - left;
                    node.y = node.y - top;
                    if (node.text) {
                        node.text.x = node.text.x - left;
                        node.text.y = node.text.y - top;
                    }
                });
                edges.forEach(function (edge) {
                    if (edge.startPoint) {
                        edge.startPoint.x = edge.startPoint.x - left;
                        edge.startPoint.y = edge.startPoint.y - top;
                    }
                    if (edge.endPoint) {
                        edge.endPoint.x = edge.endPoint.x - left;
                        edge.endPoint.y = edge.endPoint.y - top;
                    }
                    if (edge.text) {
                        edge.text.x = edge.text.x - left;
                        edge.text.y = edge.text.y - top;
                    }
                    if (edge.pointsList) {
                        edge.pointsList.forEach(function (point) {
                            point.x = point.x - left;
                            point.y = point.y - top;
                        });
                    }
                });
            }
        }
        return data;
    },
    /**
     * 显示导航
     * 显示视口范围
     * 1. 基于画布的范围比例，设置视口范围比例。宽度默认为导航宽度。
     */
    __setView: function () {
        var e_1, _a;
        // 1. 获取到图中所有的节点中的位置，将其偏移到原点开始（避免节点位置为负的时候无法展示问题）。
        var data = MiniMap.__resetData(MiniMap.__lf.getGraphRawData());
        // 由于随时都会有新节点注册进来，需要同步将注册的
        var viewMap = MiniMap.__lf.viewMap;
        var modelMap = MiniMap.__lf.graphModel.modelMap;
        var minimapViewMap = MiniMap.__lfMap.viewMap;
        try {
            // todo: no-restricted-syntax
            for (var _b = __values(viewMap.keys()), _c = _b.next(); !_c.done; _c = _b.next()) {
                var key = _c.value;
                if (!minimapViewMap.has(key)) {
                    MiniMap.__lfMap.setView(key, viewMap.get(key));
                    MiniMap.__lfMap.graphModel.modelMap.set(key, modelMap.get(key));
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        MiniMap.__lfMap.render(data);
        // 2. 将偏移后的数据渲染到minimap画布上
        // 3. 计算出所有节点在一起的边界。
        var _d = MiniMap.__getBounds(data), left = _d.left, top = _d.top, right = _d.right, bottom = _d.bottom;
        // 4. 计算所有节点的边界与minimap看板的边界的比例.
        var realWidthScale = MiniMap.__width / (right - left);
        var realHeightScale = MiniMap.__height / (bottom - top);
        // 5. 取比例最小的值，将渲染的画布缩小对应比例。
        var innerStyle = MiniMap.__miniMapWrap.firstChild.style;
        var scale = Math.min(realWidthScale, realHeightScale);
        innerStyle.transform = "matrix(" + scale + ", 0, 0, " + scale + ", 0, 0)";
        innerStyle.transformOrigin = 'left top';
        innerStyle.height = bottom - Math.min(top, 0) + "px";
        innerStyle.width = right - Math.min(left, 0) + "px";
        MiniMap.__viewPortScale = scale;
        MiniMap.__setViewPort(scale, {
            left: left,
            top: top,
            right: right,
            bottom: bottom,
        });
    },
    // 设置视口
    __setViewPort: function (scale, _a) {
        var left = _a.left, right = _a.right;
        var viewStyle = MiniMap.__viewport.style;
        viewStyle.width = MiniMap.__width - 4 + "px";
        viewStyle.height = (MiniMap.__width - 4) / (MiniMap.__lf.width / MiniMap.__lf.height) + "px";
        // top
        var _b = MiniMap.__lf.getTransform(), TRANSLATE_X = _b.TRANSLATE_X, TRANSLATE_Y = _b.TRANSLATE_Y;
        var realWidth = right - left;
        // 视口实际宽 = 视口默认宽 / (所有元素一起占据的真实宽 / 绘布宽)
        var realViewPortWidth = (MiniMap.__width - 4) / (realWidth / MiniMap.__lf.width);
        // 视口实际高 = 视口实际宽 / (绘布宽 / 绘布高)
        var realViewPortHeight = realViewPortWidth / (MiniMap.__lf.width / MiniMap.__lf.height);
        MiniMap.__viewPortTop = TRANSLATE_Y > 0 ? 0 : -TRANSLATE_Y * scale;
        MiniMap.__viewPortLeft = -TRANSLATE_X * scale;
        MiniMap.__viewPortWidth = realViewPortWidth;
        MiniMap.__viewPortHeight = realViewPortHeight;
        viewStyle.top = MiniMap.__viewPortTop + "px";
        viewStyle.left = MiniMap.__viewPortLeft + "px";
        viewStyle.width = realViewPortWidth + "px";
        viewStyle.height = realViewPortHeight + "px";
    },
    // 预览视窗
    __createViewPort: function () {
        var div = document.createElement('div');
        div.className = 'lf-minimap-viewport';
        div.addEventListener('mousedown', MiniMap.__startDrag);
        MiniMap.__viewport = div;
    },
    __startDrag: function (e) {
        document.addEventListener('mousemove', MiniMap.__drag);
        document.addEventListener('mouseup', MiniMap.__drop);
        MiniMap.__startPosition = {
            x: e.x,
            y: e.y,
        };
    },
    __drag: function (e) {
        var viewStyle = MiniMap.__viewport.style;
        MiniMap.__viewPortTop += e.y - MiniMap.__startPosition.y;
        MiniMap.__viewPortLeft += e.x - MiniMap.__startPosition.x;
        viewStyle.top = MiniMap.__viewPortTop + "px";
        viewStyle.left = MiniMap.__viewPortLeft + "px";
        MiniMap.__startPosition = {
            x: e.x,
            y: e.y,
        };
        var centerX = (MiniMap.__viewPortLeft + MiniMap.__viewPortWidth / 2)
            / MiniMap.__viewPortScale;
        var centerY = (MiniMap.__viewPortTop + MiniMap.__viewPortHeight / 2)
            / MiniMap.__viewPortScale;
        MiniMap.__lf.focusOn({
            coordinate: {
                x: centerX + MiniMap.__resetDataX,
                y: centerY + MiniMap.__resetDataY,
            },
        });
    },
    __drop: function () {
        document.removeEventListener('mousemove', MiniMap.__drag);
        document.removeEventListener('mouseup', MiniMap.__drop);
    },
};
exports.MiniMap = MiniMap;
exports.default = MiniMap;
